Add new recipe how to deploy an Inertia Rails app with Kamal and SSR support

brodienguyen opened the pull request:
This PR adds a new recipe to the Cookbook list, providing detailed steps on how to deploy an Inertia Rails application with Kamal, configured for SSR support.

![BrodieMFL 2024-12-04 at 21 24 58](https://github.com/user-attachments/assets/39b6e943-295a-456a-b403-1063a1a53f01)


> Special thanks to @skryukov for taking your time to guide me through this journey.

brodienguyen committed:
Add new recipe how to deploy your next Inertia Rails app with Kamal and SSR support

skryukov commented on docs/cookbook/deploy-with-kamal.md:
```diff
@@ -0,0 +1,222 @@
+# Deploy with `Kamal`
+
+Rails 8 will ship with [Kamal](https://kamal-deploy.org/) preconfigured as the default deployment tool.
```
This is true when you generate a new app with `rails new -j esbuild` or other builders. By default (with importmap), you'll be missing Node.js-related dependencies.

mpressen replied:
Is this comment related to this line though ?
'Cause I'm pretty sure that Kamal is the default deployment tool for rails 8, even without `-j esbuild` flag

brodienguyen replied:
@skryukov I did update my PR to highlight the difference of example Dockerfile compared to the fresh default Rails 8. The preview of the new docs have been updated in the PR description as well. Could you help me to re-review it, please?

brodienguyen committed:
Update the Dockerfile compared to the fresh default Rails 8

brodienguyen committed:
Keep sql3 instead of postgres in the sample Dockerfile to stick with default Rails setup

brodienguyen committed:
Add missing step to configure SSR URL for Inertia's Rails adapter

brodienguyen committed:
Drop code change highlighter as it doesn't take effect on Ruby code block

brodienguyen committed:
Add the code diff highlighter back in ruby snippets

ElMassimo commented on docs/cookbook/deploy-with-kamal.md:
````diff
@@ -0,0 +1,233 @@
+```ruby
+InertiaRails.configure do |config|
+  config.ssr_enabled = ViteRuby.config.ssr_build_enabled
+  config.ssr_url = "#{ViteRuby.config.protocol}://#{ViteRuby.config.host}:13714" # [!code ++]
````
For clarity, the default value of `ssr_url` was added in:

- #74

`inertia_rails` does not depend on `vite_ruby`, nor is `vite_ruby` so commonly used that it would make sense to auto-detect it and use it as a default.

In addition, I would suggest to avoid using `ViteRuby.config`, as it refers to the configuration for the _Vite **Dev** Server_, and __not__ the SSR server.

If it works, it's by coincidence.

For example, a common setup consists in running the SSR server in the same box and using plain HTTP (without SSL), while in many cases it's desirable to run the Vite dev server over SSL so that HTTP2 can be used and connections are multiplexed (loading unbundled files faster in development).

skryukov replied:
Yeah, I guess this part is a leftover from when I tried to understand what `bin/vite ssr` does (which is basically just a shortcut) 😅

I think it's better to use something like this instead: `ssr_url: ENV.fetch('INERTIA_SSR_URL', 'http://localhost:13714')` and set `INERTIA_SSR_URL` in the Kamal deployment config (any suggestions for a better name?).

brodienguyen replied:
Thanks @ElMassimo, @skryukov for our detailed discussion and suggestion. I would agree with all those valid points and prepare another PR to allow to set Inertia SSR URL via the ENV variable:
* https://github.com/inertiajs/inertia-rails/pull/172

mpressen commented on docs/cookbook/deploy-with-kamal.md:
````diff
@@ -0,0 +1,233 @@
+```ruby
+InertiaRails.configure do |config|
+  config.ssr_enabled = ViteRuby.config.ssr_build_enabled
+  config.ssr_url = "#{ViteRuby.config.protocol}://#{ViteRuby.config.host}:13714" # [!code ++]
````
Is adding this line necessary ? Without it the default seems to work just fine.

bknoles commented:
Question for the room... why do you want to use SSR?

(genuine question, no right/wrong answers, just trying to understand how people use Inertia in the real world)

mpressen commented:
> Question for the room... why do you want to use SSR?
> 
> (genuine question, no right/wrong answers, just trying to understand how people use Inertia in the real world)

The app I'm working on needs to be SEO friendly. I believe SSR is much better than CSR on that matter.

brodienguyen commented:
> Question for the room... why do you want to use SSR?
> 
> (genuine question, no right/wrong answers, just trying to understand how people use Inertia in the real world)

Ignore the benefit of SEO, in my case, purely because I would like to propose Inertia to the team where we are building and maintain a very legacy B2B software product. There are some requirements about backward compatibility on older devices or browsers with limited Javascript support with client so we have to play a bit safer on this transition.

Also since there is a lot of CRUD screen where we barely working with form and doesn't change that frequently, with SSR we can cache it a lot longer to save some computation and loading time for our user too. Without JS, at minimum they can still submit the form and get the job done rather than raising a bunch of support ticket to us (I am still not 100% sure if it is achievable with how InertiaJS form work).

brodienguyen commented:
> Question for the room... why do you want to use SSR?
> 
> (genuine question, no right/wrong answers, just trying to understand how people use Inertia in the real world)

Ignore the benefit of SEO, in my case, purely because I would like to propose Inertia to the team where we are building and maintain a very legacy B2B software product. There are some requirements about backward compatibility on older devices or browsers with limited Javascript support in the existing contract so we have to play a bit safer on this transition.

Also since there are a lot of CRUD screens where their pages are built purely with form and don't change that frequently, with SSR we can cache them a lot longer to save some computation and loading time for our user too. Without JS, at minimum they can still submit the form and get the job done rather than raising a bunch of support ticket to us (I am still not 100% sure if it is achievable with how InertiaJS form work).

brodienguyen committed:
Update the guide to reflect the recent refactor on how to set SSR URL via ENV variable

brodienguyen commented:
This PR is ready to review. I am still marking it as draft because it depends on another change at:
* https://github.com/inertiajs/inertia-rails/pull/172

skryukov commented on docs/cookbook/deploy-with-kamal.md:
````diff
@@ -207,19 +207,10 @@ export default defineConfig({
-```
+the correct SSR server URL in the adapter. It can be set via the `INERTIA_SSR_URL` ENV variable.
+
+Given we are using kamal 2, I would suggest to leverage [kamal secret](https://kamal-deploy.org/docs/commands/secrets/)
````
Why so? URL without passwords in it isn't a secret at all 😄

brodienguyen replied:
It's true. Ignore my overthinking bro 😂

brodienguyen committed:
Drop the suggestion to use kamal secret for Rails env

skryukov committed:
Merge remote-tracking branch 'origin/master' into docs-add-new-deploy-with-kamal-recipe

skryukov committed:
Modernize the Kamal recipe for the current SSR stack

Verified end-to-end on a DigitalOcean droplet with a fresh Rails 8.1 app:
both the Puma plugin path (recommended) and the separate SSR container
path deploy successfully and serve server-rendered pages.

- Recommend the built-in Puma plugin (rails 3.20+); no extra Kamal role
  needed for single-server setups
- Keep the separate-container option with a network alias and
  INERTIA_SSR_URL read by the app initializer (the gem-level env var
  from #172 was never merged)
- Build the SSR bundle via ssrBuildEnabled + ssrEntrypoint in
  config/vite.json so assets:precompile emits both bundles inside
  docker build (without ssrEntrypoint the @inertiajs/vite setup fails
  with 'No SSR entrypoint available')
- Shrink the Dockerfile section to a focused move-this-block diff so it
  does not rot with each Rails release

skryukov committed:
Note that client-rendered apps still need Node.js in the Dockerfile build stage

skryukov committed:
Bridge the whole public directory in asset_path

Inertia Rails apps output fingerprinted assets to both public/vite
(Vite Ruby) and public/assets (Propshaft); Kamal supports a single
asset_path (basecamp/kamal#967 declined arrays), so bridge their common
parent. Kamal's no-clobber merge keeps unfingerprinted files correct
per version.

skryukov committed:
Fix SSR guide gaps found while verifying the Kamal recipe

The Vite plugin path never explained how the SSR bundle gets built in
production or how the adapter gets enabled:

- Document ssrEntrypoint + ssrBuildEnabled in config/vite.json; without
  ssrEntrypoint, vite build --ssr fails with 'No SSR entrypoint
  available' (vite-plugin-ruby resolves the entry itself and the
  Inertia plugin only supplies rollup input)
- Add the missing 'Enable SSR in the adapter' step (ssr_enabled gates
  rendering in development too)
- Fix the Production example: the bundle lands in public/vite-ssr,
  not public/assets-ssr; point to the Puma plugin as the recommended
  runner
- Fix typos: 'cinfig', 'pges', and a stray 'Laravel project'

Verified against a fresh Rails 8.1 app: assets:precompile emits both
bundles and bin/vite ssr serves the bundle.

skryukov committed:
Scope SSR build instructions per Vite integration gem

The previous commit made the Vite-plugin SSR docs vite_rails-specific.
The adapter supports two integrations with different build stories:

- vite_rails (vite-plugin-ruby): needs ssrEntrypoint + ssrBuildEnabled
  in config/vite.json; bundle lands in public/vite-ssr; run with
  bin/vite ssr
- rails_vite (rails-vite-plugin + jsbundling): no extra Vite config,
  the Inertia plugin supplies the SSR entry; build via the package.json
  build script; bundle lands in ssr/; run with node ssr/ssr.js

The guide's production section now covers both, and the Kamal recipe
notes the rails_vite adjustments.

skryukov committed:
Fix SSR docs based on a four-way verification matrix

Verified all four combinations (vite_rails/rails_vite x Vite-plugin/
manual setup) with real apps, production builds, and running SSR
servers, including the Puma plugin and rails_vite's dev-mode probe
(tmp/rails-vite.json). Fixes:

- SSR bundles are NOT self-contained by default (bare imports of
  react/@inertiajs); the Kamal recipe now adds ssr.noExternal before
  recommending rm -rf node_modules, verified by booting the bundle
  with node_modules removed
- rails_vite's precompile hook builds only the client bundle and never
  runs package.json scripts; document the vite:build_ssr rake
  enhancement instead of the build-script advice
- rails_vite SSR bundle is named after the entry (ssr/inertia.js),
  not ssr/ssr.js
- ssr.entry resolves against the Vite root (source dir for vite_rails,
  project root for rails_vite); wrong paths fall back to CSR silently
- Setting ssr_url unconditionally disables dev-mode SSR (it overrides
  dev server detection); scope it to production
- Manual-setup SSR examples were missing imports in all three
  framework tabs (createInertiaApp, renderToString, createSSRApp, h,
  render); document createServer's port option and the pages:
  shorthand's dependency on @inertiajs/vite
- @inertiajs/vite is ESM-only; note the package.json type:module fix
- Puma plugin ssr/*.js glob is rails-vite-plugin's ssrOutDir default,
  not the Inertia plugin's

skryukov committed:
Make noExternal conditional on production builds

Flat ssr.noExternal breaks dev-mode SSR ('module is not defined'
evaluating the entry — CJS-only packages like React must stay external
for the Vite dev server). Use the starter kits' battle-tested form:
noExternal only when command === 'build'. Verified both modes on a
fresh app: dev SSR renders and the production bundle boots without
node_modules. Also point readers at the starter kits, which ship the
whole Kamal+SSR arrangement.

skryukov committed:
Fix leftover Laravel path in the custom SSR entry point section

skryukov committed:
Use vite-integration tabs and production-scoped ssr_enabled

- Replace the vite_rails/rails_vite prose branching with synced code
  tabs (key:vite) in the guide and the Kamal recipe
- Manual setup: advise config.ssr_enabled = Rails.env.production?
  (stack-agnostic, no dev-mode SSR noise; without the Vite plugin an
  unconditional true logs a failed SSR attempt per dev request —
  verified); keep the ViteRuby tie as a vite_rails alternative
- Reword the node_modules removal consequence: the fallback is logged,
  not silent, and on_ssr_error can report it

skryukov committed:
Advise ssr_enabled = true for manual setup

Rails.env.production? silently disables SSR in staging and other
prod-like environments, and the CSR fallback makes that invisible.
Plain true is correct everywhere a server runs; the development-mode
logged fallback is expected and documented, with the ViteRuby tie as
the opt-out for vite_rails users.
